src/depthai_wrapper.cpp

```diff
@@ -0,0 +1,169 @@
+        rgb_height_ = 3040;
+        rgb_width_ = 4056;
+        mono_height_ = 720;
+        mono_width_ = 1280;
+    }
+}
+
+void DepthAI::create_frame_holders()
+{
+
+    if (config_json_.contains("streams"))
+    {
+        for (auto it : config_json_.at("streams"))
+        {
+            // "metaout"
+            if (it.is_string())
+            {
+                if (it.get<std::string>() == "metaout")
+                    continue;
+                stream_names_.push_back(it.get<std::string>());
+                if (it.get<std::string>() == "previewout")
+                {
+                    std::shared_ptr<cv::Mat> img =
+                        std::make_shared<cv::Mat>(rgb_height_, rgb_width_, CV_8UC3);
+                    image_streams_.push_back(img);
+                }
+                else
+                {
+                    std::shared_ptr<cv::Mat> img =
+                        std::make_shared<cv::Mat>(mono_height_, mono_width_, CV_8UC3);
+                    image_streams_.push_back(img);
+                }
+            }
+            else
+            {
+                const auto& name = it.at("name").get<std::string>();
+                // streams.emplace_back(name);
+                // auto &stream = streams.back();
+                // std::cout << name << " -streams" << std::endl;
+                stream_names_.push_back(name);
+                if (name == "previewout")
+                {
+                    std::shared_ptr<cv::Mat> img =
+                        std::make_shared<cv::Mat>(rgb_height_, rgb_width_, CV_8UC3);
+                    image_streams_.push_back(img);
+                }
+                else if (name == "depth")
+                {
+                    std::shared_ptr<cv::Mat> img =
+                        std::make_shared<cv::Mat>(mono_height_, mono_width_, CV_16UC1);
+                    image_streams_.push_back(img);
+                }
+                else if (name == "disparity_color")
+                {
+                    std::shared_ptr<cv::Mat> img =
+                        std::make_shared<cv::Mat>(mono_height_, mono_width_, CV_16UC1);
+                    image_streams_.push_back(img);
+                }
+            }
+        }
+    }
+}
+
+DepthAI::~DepthAI() = default;
+
+// How about creating streams for everything in the vector in constructor and then updating them
+// each time get frames is called ?
+
+// TODO: Crash on size mismatch. DO this later.
+// TODO: save config as json and check if the streams are in the config and crash if they aren't
+
+void DepthAI::get_frames(std::unordered_map<std::string, CV_mat_ptr>& output_streams)
+{
+    // std::cout << "mono it-----" << mono_height_ << " ~~ " << mono_width_ << std::endl;
+
+    int count = stream_names_.size();
+    std::vector<std::string>::iterator it;
+    std::vector<bool> dirty_check(stream_names_.size(), false);
+    while (count)
+    {
+        packets_ = pipeline_->getAvailableNNetAndDataPackets(true);
```

Comment:
Is there a reason why `packets_` is a class variable? It could just be a local variable here.

Reply:
it can be a local.

Reply:
doesn't that make an allocation and dealocation of memory each time it's called. doesn't avoiding that is better ? I am not sure what do you think ?

Reply:
General rule of thumb -- always prefer local variables over globals. Here, there's no need for `packets_` to be a class variable since it is overwriiten many times inside the loop. Keeping it local will only need for it to be allocated on the stack on each function call. Compiler will optimize popping and pushing on/off the stack. But if you create it a class variable, it needs to be stored in the symbol table and then resolved, which is not needed here at all.